ocular/src/chain/client/transactions.rs

```diff
@@ -0,0 +1,170 @@
+use crate::error::TransactionError;
+use cosmrs::{
+    bank::MsgSend,
+    crypto::{secp256k1::SigningKey, PublicKey},
+    staking::{MsgDelegate, MsgUndelegate},
+    tendermint::chain::Id,
+    tx::{self, Fee, Msg, SignDoc, SignerInfo},
+    AccountId, Coin,
+};
+use tendermint_rpc::endpoint::broadcast::tx_commit::Response;
+
+use super::ChainClient;
+
+/// Metadata wrapper for transactions
+#[derive(Clone, Debug)]
+pub struct TransactionMetadata {
+    pub chain_id: Id,
+    pub account_number: u64,
+    pub sequence_number: u64,
+    pub gas_limit: u64,
+    pub timeout_height: u16,
+    pub memo: String,
+}
+
+impl ChainClient {
+    // TODO: Make this extensible to multisig and multicoin (or add new methods for that)
+    /// Helper method for signing and broadcasting messages.
+    async fn sign_and_send_msg(
+        &self,
+        sender_public_key: PublicKey,
+        sender_private_key: SigningKey,
+        amount: Coin,
+        tx_body: tx::Body,
+        tx_metadata: TransactionMetadata,
+    ) -> Result<Response, TransactionError> {
+        // Create signer info.
+        let signer_info =
+            SignerInfo::single_direct(Some(sender_public_key), tx_metadata.sequence_number);
+
+        // Compute auth info from signer info by associating a fee.
+        // TODO: Add ability to specify payers and granters for fees.
+        let auth_info =
+            signer_info.auth_info(Fee::from_amount_and_gas(amount, tx_metadata.gas_limit));
+
+        // Create doc to be signed
+        let sign_doc = match SignDoc::new(
+            &tx_body,
+            &auth_info,
+            &tx_metadata.chain_id,
+            tx_metadata.account_number,
+        ) {
+            Ok(doc) => doc,
+            Err(err) => return Err(TransactionError::TypeConversionError(err.to_string())),
+        };
+
+        // Create raw signed transaction.
+        let tx_signed = match sign_doc.sign(&sender_private_key) {
+            Ok(raw) => raw,
+            Err(err) => return Err(TransactionError::SigningError(err.to_string())),
+        };
+
+        // Broadcast transaction
+        match tx_signed.broadcast_commit(&self.rpc_client).await {
+            Ok(response) => Ok(response),
+            Err(err) => Err(TransactionError::BroadcastError(err.to_string())),
+        }
+    }
+
+    // TODO: Make this extensible to multisig and multicoin (or add new methods for that)
+    /// Signs and sends a simple transaction message.
+    pub async fn sign_and_send_msg_send(
+        &self,
+        sender_account: AccountId,
+        sender_public_key: PublicKey,
+        sender_private_key: SigningKey,
```

Comment:
Maybe we can think of a way to set the account to be used by the client ahead of time so that all of this information is implicit in any transaction calls so the user doesn't need to pass them separately? Maybe this means having a higher level `Account` type that includes the `AccountId`, `PublicKey` and `SigningKey`?
This would allow the client to be configured to run with a specific account and make the amount of code needed to send a tx much shorter. Thoughts?

Reply:
Hm. We can definitely have an account type that includes all of the above. However, that would only be for the sender (since it's possible, if not likely, that the user will not have recipient account public or private keys).

Insofar as having the client configured to run with a specific account, I think maybe it might be best to give the user the most autonomy possible and not limit them to just 1 account per client. There's likely a few different use cases where the user would want to do different operations across different accounts within 1 client.

Reply:
Yeah I think we just keep the recipient to an address like it is now.

I thought about that, we should think about the best way to configure the client to run cleanly with a set account but also support other stuff.

Reply:
Updated to include a new account type.

Regarding the set/default account, I'm not super sure what the best way to manage that would be. We could add in a 'default' account to each chain client, however the one caveat/consideration there would be if actual default value vs expected default value differ for some reason. Worst case scenario would be a user performing a transaction with a different account than intended bc they're relying on a default value as opposed to them having to explicitly pass it to perform a transaction. That being said I'm still open to default values if we think the UX would improve significantly.

Reply:
I'm definitely still thinking through the UX of this and it may take some time so I don't want you to keep making code changes after my comments on it. We could merge and change later.

With that said, maybe a method-chain approach could be nice? like

```
client.send_to(address)
    .from(account)
    // ... more chaining options for other metadata
```

Reply:
Hmm. I personally like longer well defined methods (even if we'll require more of them to support various presets/defaults and whatnot) as opposed to method chaining just b/c it feels more clear and explicit (to me personally).

That might just be my object-oriented background talking though :) maybe we could get a few more opinions and see what we can get quorum on.

Reply:
yeah I definitely think it's a design question that i'd like more input on. @tony-iqlusion @jackzampolin @zmanian